### ISSUE TITLE
[Bug Report] Low PCC of `ttnn.global_avg_pool2d` after running `ttnn.max_pool2d`

### ISSUE THREAD

**Issue by jerrysky3** [tenstorrent/tt-metal#15824]
**Describe the bug**
The PCC of `ttnn.global_avg_pool2d` becomes very low (0.7) if we run `ttnn.max_pool2d` first for some shapes

**To Reproduce**

Run the example below:

```py
import torch
import ttnn
from tests.ttnn.utils_for_testing import assert_with_pcc


def test_maxpool(device, input_shape, kernel_size, stride, padding, dilation):
    torch_input = torch.rand(input_shape, dtype=torch.bfloat16)
    batch_size, in_c, in_h, in_w = input_shape

    input_tensor = torch.permute(torch_input, (0, 2, 3, 1))
    input_tensor = torch.reshape(input_tensor, (1, 1, -1, in_c))
    input_tensor = ttnn.from_torch(input_tensor, layout=ttnn.ROW_MAJOR_LAYOUT, device=device)
    output_tensor = ttnn.max_pool2d(
        input_tensor,
        batch_size,
        in_h,
        in_w,
        in_c,
        kernel_size,
        stride,
        padding,
        dilation,
    )

    expected_output = torch.nn.functional.max_pool2d(torch_input, kernel_size, stride, padding)

    output_tensor = ttnn.to_torch(output_tensor)
    _, out_c, out_h, out_w = expected_output.shape
    output_tensor = torch.reshape(output_tensor, (batch_size, out_h, out_w, out_c))
    output_tensor = torch.permute(output_tensor, (0, 3, 1, 2))

    assert torch.allclose(output_tensor, expected_output), "mismatch"


def test_global_avg_pool(device, input_shape):
    torch_input_tensor = torch.randn(input_shape, dtype=torch.bfloat16)
    torch_output_tensor = torch.nn.functional.adaptive_avg_pool2d(torch_input_tensor, (1, 1))

    input_tensor = torch.permute(torch_input_tensor, (0, 2, 3, 1))
    input_tensor = ttnn.from_torch(input_tensor, layout=ttnn.TILE_LAYOUT, device=device)
    output_tensor = ttnn.global_avg_pool2d(input_tensor)
    output_tensor = ttnn.to_torch(output_tensor)
    output_tensor = torch.permute(output_tensor, (0, 3, 1, 2))

    assert_with_pcc(torch_output_tensor, output_tensor)


if __name__ == "__main__":
    try:
        device = ttnn.open_device(device_id=0, l1_small_size=4096)
        test_maxpool(device, (1, 192, 56, 56), (2, 2), (2, 2), (0, 0), (1, 1))
        test_global_avg_pool(device, (1, 64, 1, 32))
    finally:
        ttnn.close_device(device)
```

**Expected behavior**
The PCC check of `ttnn.global_avg_pool2d` should pass (> 0.9999)

**Please complete the following environment information:**
- OS: Ubuntu 20.04
- Version of software  612744d7e9d8810f20c1458710c7f5eab361bb7e
- Card: N300

**Additional context**
The PCC issue of `ttnn.global_avg_pool2d` seems to be persistent once `ttnn.max_pool2d` is run, unless the card is reset

**Comment by mywoodstock** [comment]
Some observations during debug:

Result of `global_avg_pool` (SUM reduce) seems bit ND. Most of the time it is a deterministic value, but varies at times. Here is a sequence of 20 runs of `maxpool` followed by `global_avg_pool`:
```
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.682736883289193
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.682736883289193
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.682736883289193
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.9999999992819725
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.682736883289193
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.682736883289193
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.682736883289193
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.9999999992819725
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.682736883289193
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.682736883289193
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.682736883289193
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.9999999992819725
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.682736883289193
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.682736883289193
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.682736883289193
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.9999999992819725
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.682736883289193
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.682736883289193
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.682736883289193
MAXPOOL PCC: 1.0
AVGPOOL PCC: 0.9999999992819725
```
The pattern repeats for as many runs as I try, where first 3 runs have bad PCC, and every 4th has good.

I changed to directly use reduce op (mean), instead of avg_pool, and I see exact same thing.

In the output tensor from the reduce (mean) op, I see that of the 64 elements, first 16 are 0s (wrong) followed by 16 correct values, then another 16 0s, followed by another 16 correct values.

My suspicion is that some configuration register might be getting messed up. I tried using the `tilize_uninit` and `pack_untilize_uninit`, but those didn't help. Since we are using `tilizeA_B_reduce_init`, perhaps we need an LLK uninit corresponding to it?

It would be good if someone in LLK team could take a look at this. I am using the following code as the test:

```
import torch
import ttnn
from tests.ttnn.utils_for_testing import assert_with_pcc, check_with_pcc


def test_maxpool(device, input_shape, kernel_size, stride, padding, dilation):
    torch.manual_seed(0)
    torch_input = torch.rand(input_shape, dtype=torch.bfloat16)
    batch_size, in_c, in_h, in_w = input_shape

    input_tensor = torch.permute(torch_input, (0, 2, 3, 1))
    input_tensor = torch.reshape(input_tensor, (1, 1, -1, in_c))
    input_tensor = ttnn.from_torch(input_tensor, layout=ttnn.ROW_MAJOR_LAYOUT, device=device)
    output_tensor = ttnn.max_pool2d(
        input_tensor,
        batch_size,
        in_h,
        in_w,
        in_c,
        kernel_size,
        stride,
        padding,
        dilation,
    )

    expected_output = torch.nn.functional.max_pool2d(torch_input, kernel_size, stride, padding)

    output_tensor = ttnn.to_torch(output_tensor)
    _, out_c, out_h, out_w = expected_output.shape
    output_tensor = torch.reshape(output_tensor, (batch_size, out_h, out_w, out_c))
    output_tensor = torch.permute(output_tensor, (0, 3, 1, 2))

    assert torch.allclose(output_tensor, expected_output), "mismatch"
    pcc, msg = check_with_pcc(expected_output, output_tensor)
    print(f"MAXPOOL PCC: {msg}")
    assert pcc


def test_global_avg_pool(device, input_shape):
    torch.manual_seed(0)
    torch_input_tensor = torch.randn(input_shape, dtype=torch.bfloat16)
    # torch_output_tensor = torch.nn.functional.adaptive_avg_pool2d(torch_input_tensor, (1, 1))
    torch_output_tensor = torch.mean(torch_input_tensor, dim=-1, keepdim=True)

    input_tensor = torch.permute(torch_input_tensor, (0, 2, 3, 1))
    input_tensor = ttnn.from_torch(input_tensor, layout=ttnn.TILE_LAYOUT, device=device)
    # output_tensor = ttnn.global_avg_pool2d(input_tensor)
    output_tensor = ttnn.mean(input_tensor, dim=-2, keepdim=True)
    output_tensor = ttnn.to_torch(output_tensor)
    output_tensor = torch.permute(output_tensor, (0, 3, 1, 2))
    
    # torch.save(torch_output_tensor, "torch_output.pt")
    # torch.save(output_tensor, "output.pt")

    # assert_with_pcc(torch_output_tensor, output_tensor)
    pcc, msg = check_with_pcc(torch_output_tensor, output_tensor)
    print(f"AVGPOOL PCC: {msg}")
    assert pcc


if __name__ == "__main__":
    try:
        device = ttnn.open_device(device_id=0, l1_small_size=4096)
        test_maxpool(device, (1, 192, 56, 56), (2, 2), (2, 2), (0, 0), (1, 1))
        test_global_avg_pool(device, (1, 64, 1, 32))
    finally:
        ttnn.close_device(device)
```

**Comment by mywoodstock** [comment]
@ttmtrajkovic @rtawfik01 could someone from LLK please take a look if there is anything missing with init/uninit?

**Comment by mywoodstock** [comment]
Relevant kernels:

maxpool compute: `ttnn/cpp/ttnn/operations/pool/generic/device/kernels/compute/max_pool_multi_core.cpp`
reduce compute: `ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/compute/reduce_h.cpp`

**Comment by mywoodstock** [comment]
Another signal is that this issue happens when the maxpool kernel size is either `(1, 1)` or `(2, 2)`, with `(3, 3)` and above output looks good always. The kernel size volume (1*1, or 2*2, etc) is used to configure the reduce function to reduce those many rows in a face, so may be some configuration gets stuck?

**Comment by bbradelTT** [comment]
I adjusted the example to just call reduce directly.

https://github.com/tenstorrent/tt-metal/issues/6991#issuecomment-2559949342

Found the following:

I tried making changes to ttnn/cpp/ttnn/operations/pool/generic/device/kernels/compute/max_pool_multi_core.cpp

After adding

    tilize_uninit(in_cb_id,out_cb_id);
    pack_untilize_uninit(out_cb_id);
the PCC is still bad on the second call.

After commenting out

    pack_untilize_dst_init_short<in_ntiles_c>(
        out_cb_id, num_out_rows, num_faces_in_tile); /* pack 1 row (1x16 or 1x32) */
the PCC is good on the second call.

cc @ncvetkovicTT

**Comment by bbradelTT** [comment]
Narrowed it down further:

Commented out

tt_metal/include/compute_kernel_api/pack_untilize.h

    PACK((llk_pack_untilize_init<block_ct_dim, full_ct_dim, diagonal, narrow_row, row_num_datums>(
        ocb, face_r_dim, num_faces)));
and reduce/test_global_avg_pool worked.

**Comment by bbradelTT** [comment]
@ncvetkovicTT 

Despite the name, the file is a regular program (that has a main section), not a test program.
You need to run with python and not with pytest.

I provided a full example of how to run in my issue: https://github.com/tenstorrent/tt-metal/issues/6991#issuecomment-2559885872

**Comment by ncvetkovicTT** [comment]
@bbradelTT No update unfortunately, I suspect that y_cnt for either input or output on PACK side is not correctly set. To verify this, it would maybe be helpful to change ```test_mean``` to something really simple, like a single tile with arranged values and instead of mean do copy to cb_out only. I tried doing that, but I wasn't successful, could you perhaps provide me an example that is the same as the one above, just instead of mean do copy? If the issue is that only reduce after max_pool fails, then I'll resort to simulation.

**Comment by ncvetkovicTT** [comment]
A bit more detail: This issue is being tracked together with #6991. They happen only on WH. In the example above there's the same pattern as explained in #6991, where max_pool+mean passes on every 4th attempt. In order to see if the pattern is related to mean, or irrelevant of the op, we should try with a different op after max_pool.

**Comment by nkpatel-tt** [comment]
> More context: This issue is being tracked together with [#6991](https://github.com/tenstorrent/tt-metal/issues/6991). They happen only on WH. In the example from the mentioned issue there's the same pattern as explained in [#6991](https://github.com/tenstorrent/tt-metal/issues/6991), where max_pool+mean passes on every 4th attempt. In order to see if the pattern is related to mean, or irrelevant of the op, we should try with a different op after max_pool.

One bug reported by while enabling max pool. It seems this is also in same line.
https://github.com/tenstorrent/tt-metal/issues/15824

**Comment by ncvetkovicTT** [comment]
@bbradelTT If I make ```input_shape``` for ```test_maxpool``` to be (1, 192, 32, 32) and not (1, 192, 56, 56), the test passes always. If I lower 3rd dimension down further to 64, the test passes, but for (1, 32, 32, 32) it passes every 2nd, not every 4th time. 

Could you explain a bit what goes on in ```test_maxpool```? I can't figure out how padding works, what actually gets permutet and what is expected input tensor to end up in L1. This is maybe important because whenever we have some LLKs with flags for narrow tiles or partial faces, they end up being false, which leads me to believe that the input is padded, but maybe poorly, because whenever the input shape is (1, C*32, H*32, W*32) the test passes (values are random, with a fixed seed).

**Comment by bbradelTT** [comment]
@ncvetkovicTT in regards to what this is related to, I think it's related to `reduce_init`/`reduce_tile`

For the basic reduce unit tests`ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/compute/reduce_w.cpp` uses `mm_init`/`matmul_tiles`. When that happens, when run after max_pool, the tests pass. However, when I commented out that option and forced `reduce_init`/`reduce_tile` to be used, I saw the same behaviour.

That is:
1. max_pool + reduce_w using matmul - passes on the first attempt
2. max_pool + reduce_w using reduce - fails on the first attempt (I did not try 4+ attempts, although I'm guessing the same pattern would exist)

You should be able to test via the code in https://github.com/tenstorrent/tt-metal/issues/6991#issuecomment-2559885872 by changing
```
test_mean(device, 1, 32, 32, -2)
```
to
```
test_mean(device, 1, 32, 32, -1)
```

In terms of what else uses reduce_tile, running `grep 'reduce_tile(' -r *` in tt-metal shows other ops use reduce_tile as well. E.g. layernorm, groupnorm, and softmax. There are also a lot of tests:
```
tests/tt_metal/tt_metal/test_kernels/compute/layernorm.cpp:                reduce_tile(cb_x, cb_scaler, wt + j, scaler0, dst0);
tests/tt_metal/tt_metal/test_kernels/compute/layernorm.cpp:                reduce_tile(cb_xmm2, cb_scaler, wt + wtr, scaler0, dst0);
tests/tt_metal/tt_metal/test_kernels/compute/layernorm.cpp:            // reduce_tile(cb_xmm, cb_scaler, wt+wtr, scaler0, dst0);
tests/tt_metal/tt_metal/test_kernels/compute/max_pool.cpp:            reduce_tile(in_cb_id, in_scalar_cb_id, tile_i, 0, dst_i);
tests/tt_metal/tt_metal/test_kernels/compute/rmsnorm.cpp:                reduce_tile(cb_x2, cb_scaler, wt + wtr, scaler0, dst0);
tests/tt_metal/tt_metal/test_kernels/compute/rmsnorm.cpp:            // reduce_tile(cb_xmm, cb_scaler, wt+wtr, scaler0, dst0);
tests/tt_metal/tt_metal/test_kernels/compute/max_pool_multi_core.cpp:            reduce_tile(in_cb_id, in_scalar_cb_id, tile_i, 0, dst_i);
tests/tt_metal/tt_metal/test_kernels/compute/reduce_h.cpp:                reduce_tile(tt::CBIndex::c_0, tt::CBIndex::c_2, 0, 0, reduce_dst_idx);
tests/tt_metal/tt_metal/test_kernels/compute/reduce_hw.cpp:                reduce_tile(tt::CBIndex::c_0, tt::CBIndex::c_2, 0, 0, reduce_dst_idx);
tests/tt_metal/tt_metal/test_kernels/compute/softmax.cpp:            reduce_tile(cb_exps, cb_bcast_scaler, wt, bcast_scaler0, dst0);
tests/tt_metal/tt_metal/test_kernels/compute/reduce_w.cpp:                reduce_tile(tt::CBIndex::c_0, tt::CBIndex::c_2, 0, 0, reduce_dst_idx);
```
Trying this other stuff out might be a big time sink though.

**Comment by bbradelTT** [comment]
@ncvetkovicTT maxpool is @mywoodstock 's area of expertise. I don't know the details.

**Comment by bbradelTT** [comment]
56*4=224=7*32

It makes sense that this works every fourth iteration.

**Comment by nkpatel-tt** [comment]
> [@bbradelTT](https://github.com/bbradelTT) If I make `input_shape` for `test_maxpool` to be (1, 192, 32, 32) and not (1, 192, 56, 56), the test passes always. If I lower 3rd dimension down further to 64, the test passes, but for (1, 32, 32, 32) it passes every 2nd, not every 4th time.
> 
> Could you explain a bit what goes on in `test_maxpool`? I can't figure out how padding works, what actually gets permutet and what is expected input tensor to end up in L1. This is maybe important because whenever we have some LLKs with flags for narrow tiles or partial faces, they end up being false, which leads me to believe that the input is padded, but maybe poorly, because whenever the input shape is (1, C_32, H_32, W*32) the test passes (values are random, with a fixed seed).

The algorithm works as follows.
1) Reader kernel reads data from halo output of input tensor to be processed by compute kernel in height direction(`tt::tt_metal::ReduceOpDim::H`). For example, for 3x3 kernel, each output row needs 9 input rows. Reader kernel creates these rows. Here note that width dimension of buffer to be processed by compute kernel is input channels.
2) Compute kernel simply takes this buffer and does reduction using `reduce_tile_math`, kernel(`ttnn/cpp/ttnn/operations/pool/generic/device/kernels/compute/max_pool_multi_core.cpp`). 
3) Once reduction is done compute kernel writes data to output buffer using `pack_untilize_dst<num_output_tiles>`.


Partial tiles are taken care in compute kernel. However, I think that condition won't apply here since the input channels are greater than 32.
Pls do let me know if any further clarification is required.

**Comment by ncvetkovicTT** [comment]
@nkpatel-tt Thank you for trying to explain, but I am not sure I fully follow. If I put input_shape to be (1, 192, 56, 56), in compute kernel I get:

in_ntiles_hw = 1
in_ntiles_c = 6
window_size_hw = 4
out_h = 28
out_w = 28
nsticks_per_core = 13

but for (1, 192, 64, 64) I get:

in_ntiles_hw = 1
in_ntiles_c = 6
window_size_hw = 4
out_h = 32
out_w = 32
nsticks_per_core = 16

I can't see the difference in parameters between the two shapes in LLKs, that would help me understand what gets configured poorly.

**Comment by nkpatel-tt** [comment]
> [@nkpatel-tt](https://github.com/nkpatel-tt) Thank you for trying to explain, but I am not sure I fully follow. If I put input_shape to be (1, 192, 56, 56), in compute kernel I get:
> 
> in_ntiles_hw = 1 in_ntiles_c = 6 window_size_hw = 4 out_h = 28 out_w = 28 nsticks_per_core = 13
> 
> but for (1, 192, 64, 64) I get:
> 
> in_ntiles_hw = 1 in_ntiles_c = 6 window_size_hw = 4 out_h = 32 out_w = 32 nsticks_per_core = 16
> 
> I can't see the difference in parameters between the two shapes in LLKs, that would help me understand what gets configured poorly.

Yes that is correct. Let me elaborate

-  In each iteration of for loop (`for (uint32_t i = 0; i < nsticks_per_core; ++i)`) at line 114 in [compute kernel](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/pool/generic/device/kernels/compute/max_pool_multi_core.cpp) of max pool, one output stick is calculated. One output stick is `1 X input_channels`.
- The buffer passed to `unpack_tilizeA_B_block` at line no. 61 in [compute kernel](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/pool/generic/device/kernels/compute/max_pool_multi_core.cpp)  contains kernel_h X kernel_w rows of input. The output expected is 1 row.
-  So each core is given several outputs to find max_pool of and reduce operation is called on them one after another.
-  The configurations can only change in case there is change in input channels, kernel size. In your prints above, in_ntiles_c(`6*32 = 192`) and window_size_hw(`kernel_h*kernel_w`) are relevant for config change

I did several experiments with example in this [thread](https://github.com/tenstorrent/tt-metal/issues/15824#issuecomment-2555973531) and by borys [here](https://github.com/tenstorrent/tt-metal/issues/6991#issuecomment-2559885872) and for all of the following configurations the result is same, every 4th iteration the PCC is good.

- test_maxpool(device, (1, 192, 56, 56), (2, 2), (2, 2), (0, 0), (1, 1)
- test_maxpool(device, (1, 192, 64, 64), (2, 2), (2, 2), (0, 0), (1, 1))
- test_maxpool(device, (1, 192, 32, 32), (2, 2), (2, 2), (0, 0), (1, 1))
- test_maxpool(device, (1, 64, 32, 32), (2, 2), (2, 2), (0, 0), (1, 1))

My set-up

- latest main(2a86ff7fdde8561ec6c60f1635937ac2b13e122f)
- Board N300 (ran `export WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml` before doing experiment),

**Comment by ayerofieiev-tt** [comment]
Raising to P0 to ensure this is resolved and not delayed further. Thank you!

**Comment by cmaryanTT** [comment]
@ncvetkovicTT  to talk to @mywoodstock to debug the issue.

**Comment by ncvetkovicTT** [comment]
@mywoodstock @cmaryanTT @bbradelTT has gone again through max_pool op for me, so that I can understand the underlying HLK better.

During the day I have tried dumping the PACK configuration which contains the info on the programmed strides for the counters and such, but I wasn't able to see the difference in the configuration when calling/not calling max_pool (I dumped PACK state from reduce_h.cpp). 

Abhinav and I have agreed on the following:
- He will try to make a single core example that uses max_pool, which may turn out to be a non-trivial endeavor. In case he succeeds I can simulate the code.

- I will once again take a look at PACK configuration, now that I fully understand what max_pool HLK does.

**Comment by ncvetkovicTT** [comment]
I was able to make the files necessary for the simulation but unable to see the signals due to some access constraints. @mywoodstock @nkpatel-tt can you confirm that if I give an input [1, 32, 2, 2] to max_pool, with the kernel [2, 2], that only one core will be utilized and that this is a valid input from the op standpoint?

**Comment by mywoodstock** [comment]
@ncvetkovicTT I believe with this input size you also need to set the stride to 2,2 so that the output size is just 1 stick.

**Comment by ncvetkovicTT** [comment]
@mywoodstock Thanks! Could you please point me to the part of the test from above where I can set that?

**Comment by mywoodstock** [comment]
That would be:
```
        test_maxpool(device, (1, 32, 2, 2), (2, 2), (2, 2), (0, 0), (1, 1))
```

**Comment by ncvetkovicTT** [comment]
TL;DR: I have tried many different things yesterday in the attempt to obtain valid simulation waveforms but wasn't successful. I will try rebasing with some older UMD changes and if that's unsuccessful raise the issue to get the simulator working.

1. When I run the simulation with both ```max_pool``` and ```mean```, I obtain the .vcd file needed for viewing the waves, but then we saw some NaN and -inf values (0xFFFF and 0xFF80, respectively) as an input to ```max_pool```.
2. Thinking that there's an issue with running both tests on the simulator, I reduced it to one (e.g. just ```test_maxpool```) and printed the output of the silicon vs. the output of the simulator. In this case, simulator returns NaN, which I see in the output of the test. 
3. Doubting the python tests, I ran some simple C++ test, namely single-core single-tile eltwise binary add, i.e. ```TensixBinaryComputeSingleCoreSingleTileAdd```. This test passes on silicon, but fails on the simulator, again returning NaN. 
4. After that I suspected that there might only be the issue with data-movement part of the simulator, so I bypassed reader kernel by hardcoding the values in L1. Silicon behaved as expected, but again the simulator returned NaN. I repeated the same thing with the writer kernel, hardcoding some value to output CB, yet again NaN from the simulator.
5. Last thing that I tried was cloning versim repo again in case I didn't build it properly. Here I saw the following warning, which I couldn't find anywhere in Slack, but anyway it's related to GS, not WH_B0:

```
ninja versim -C build
ninja: Entering directory `build'
[94/95] Building CXX object versim/CMakeFiles/versim.dir/tt_versim_device.cpp.o
In file included from ../versim/tt_versim_device.cpp:14:
In file included from ../third_party/versim/versim/wormhole_b0/headers/src/t6ifc/versim-core/sim_interactive.h:7:
In file included from ../third_party/versim/versim/wormhole_b0/headers/src/t6ifc/common/device.h:19:
In file included from ../third_party/versim/versim/wormhole_b0/headers/src/software/command_assembler/inc/command_assembler/tile_metadata.h:6:
In file included from ../third_party/versim/versim/wormhole_b0/headers/src/software/command_assembler/inc/command_assembler/dimensions.h:10:
In file included from ../third_party/versim/versim/grayskull/headers/usr_include/boost/archive/text_iarchive.hpp:24:
In file included from ../third_party/versim/versim/grayskull/headers/usr_include/boost/archive/basic_text_iarchive.hpp:30:
In file included from ../third_party/versim/versim/grayskull/headers/usr_include/boost/archive/detail/common_iarchive.hpp:23:
In file included from ../third_party/versim/versim/grayskull/headers/usr_include/boost/archive/detail/interface_iarchive.hpp:22:
../third_party/versim/versim/grayskull/headers/usr_include/boost/archive/detail/iserializer.hpp:69:7: warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if ! DONT_USE_HAS_NEW_OPERATOR
      ^
../third_party/versim/versim/grayskull/headers/usr_include/boost/archive/detail/iserializer.hpp:63:12: note: expanded from macro 'DONT_USE_HAS_NEW_OPERATOR'
        || defined(__SUNPRO_CC) && (__SUNPRO_CC < 0x590)   \
           ^
../third_party/versim/versim/grayskull/headers/usr_include/boost/archive/detail/iserializer.hpp:212:9: warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
    #if DONT_USE_HAS_NEW_OPERATOR
        ^
../third_party/versim/versim/grayskull/headers/usr_include/boost/archive/detail/iserializer.hpp:63:12: note: expanded from macro 'DONT_USE_HAS_NEW_OPERATOR'
        || defined(__SUNPRO_CC) && (__SUNPRO_CC < 0x590)   \
           ^
2 warnings generated.
[95/95] Linking CXX executable versim/versim
```

As for today, I will try switching to [Vincent's branch](https://github.com/tenstorrent/tt-metal/compare/main...vtangTT/metal_simulator_branch), introduce my changes and see if I can make the simulation work. The expectation here is that there's a discrepancy between the simulator and the UMD (since UMD is actively developed), so maybe the older version of tt-metal might get it to work.

**Comment by ncvetkovicTT** [comment]
I ran eltwise ADD C++ test on Vincent's branch, it works on silicon but doesn't on simulator, breaks with the following error:
```
libc++abi: terminating due to uncaught exception of type std::runtime_error: TT_ASSERT @ ../tt_metal/third_party/umd/device/simulation/tt_simulation_device.cpp:82: cmd == DEVICE_COMMAND_EXIT
info:
Did not receive expected command from remote.
backtrace:
 --- tt_SimulationDevice::start_device(tt_device_params const&)
 --- tt::Cluster::initialize_device_drivers()
 --- tt::Cluster::Cluster()
 --- tt::Cluster::instance()
 --- tt::tt_metal::v0::GetNumAvailableDevices()
 --- ./build/test/tt_metal/unit_tests(+0x4d3ed) [0x55aa221fb3ed]
 --- ./build/test/tt_metal/unit_tests(+0x69a79) [0x55aa22217a79]
 --- testing::internal::UnitTestImpl::PostFlagParsingInit()
 --- void testing::internal::InitGoogleTestImpl<char>(int*, char**)
 --- /localdev/ncvetkovic/tt-metal/build_Release/lib/libgtest_main.so.1.13.0(main+0x2e) [0x7f20012a016e]
 --- /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7f2000d05083]
 --- ./build/test/tt_metal/unit_tests(+0x4cc6e) [0x55aa221fac6e]

Aborted (core dumped)
```
Any ideas @vtangTT how I can continue? Ideally I would like to make the simulator work with my branch:
[ncvetkovic/15824-max_pool_and_reduce_fail](https://github.com/tenstorrent/tt-metal/tree/ncvetkovic/15824-max_pool_and_reduce_fail)

**Comment by ncvetkovicTT** [comment]
@vtangTT @ttmtrajkovic More context on how I set up Versim and my repo.

1. I did everything on bgd-lab-11 (because the [instructions](https://yyz-gitlab.local.tenstorrent.com/tenstorrent/tt-metal/-/blob/main/versim/README.md) say that I can use a machine with a card also).
2. My tt-metal branch is [ncvetkovic/15824-max_pool_and_reduce_fail](https://github.com/tenstorrent/tt-metal/tree/ncvetkovic/15824-max_pool_and_reduce_fail), there I built metal like always. The location of this repo is ```/localdev/ncvetkovic/tt-metal```
3. I cloned Versim to a separate directory ```/localdev/ncvetkovic/tt-metal-sim/tt-metal```. I followed the steps from the instructions mentioned in 2. and did everything successfully (it only dumped a warning that affects Grayskull, but I am working on Wormhole_b0).
4. I go back to tt-metal repo and run a simple eltwise_add C++ test: ```TT_METAL_SLOW_DISPATCH_MODE=1 ./build/test/tt_metal/unit_tests_llk --gtest_filter=*BinaryComputeSingleCoreSingleTileAdd```.
5. The test passes when run on silicon.
6. I then set `export TT_METAL_SIMULATOR_EN=1` and `export TT_REMOTE_EXE=/localdev/ncvetkovic/tt-metal-sim/tt-metal/scripts/versim_run.sh`. I run the same test again. The output that I get this time in the terminal is:
```
                   Test | ERROR    | Discrepacy: A=-nan, B=0.578125
                   Test | ERROR    |    absdiff=nan, atol=0.001
                   Test | ERROR    |    reldiff=nan, rtol=0.0155
../tests/tt_metal/tt_metal/llk/test_single_core_binary_compute.cpp:289: Failure
Value of: unit_tests::compute::binary::single_core_binary(devices_.at(id), test_config)
  Actual: false
Expected: true
```
Note that A=-nan. I get a .log and .vcd file, but logs always end with the following message, I don't know if that's safe to ignore:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Received unknown command from host
/localdev/ncvetkovic/tt-metal-sim/tt-metal/scripts/versim_run.sh: line 17: 396829 Aborted   
```

**Comment by ncvetkovicTT** [comment]
Hey @mywoodstock, as you can see, the workaround is merged already and it passed the CI. I propose one of the two actions:
1. Close the issue, in which case we'll open a separate tracker for finding a proper LLK fix.
2. Lower the prio for this issue and redefine it's purpose to be a tracker for the LLK fix.

**Comment by mywoodstock** [comment]
Hi @ncvetkovicTT , thanks. So the workaround works for the primary kernel. There is one more maxpool kernel for larger cases, where it performs reductions in groups of 8 tiles at a time (and remainder in the last iter). I think this issue might exist there as well. This PR: https://github.com/tenstorrent/tt-metal/pull/16544. @wransom-TT tried removing the template param for this large kernel, but is getting PCC errors, so I have suggested to keep the template parameters in the large kernel. Do you know if there is some other workaround for this case?

Yeah, we should open a new issue to get a proper LLK fix.

**Comment by ncvetkovicTT** [comment]
@mywoodstock Hey so I didn't quite get it, is there a detected PCC error similar to the one from this issue, or you guys assume it might happen there as well? I'll create a new issue for tracking the LLK bug in the morning, do I close this one then?

@amahmudTT Could you also take a quick look at #16544? Does the name or the topic ring a bell? Thought it might be related to something you worked on recently, but maybe I'm wrong.

**Comment by mywoodstock** [comment]
@ncvetkovicTT Yes, we suspect there may be same issue with the other kernel, but don't have a repro yet. SO yes lets close this issue and open a new one to track the proper LLK fix. Thanks!

**Comment by ayerofieiev-tt** [comment]
@mywoodstock , @ncvetkovicTT please make sure to post the new ticket here

**Comment by ncvetkovicTT** [comment]
The following GH issue tracks the LLK pack_untilize problem tracker:
https://github.com/tenstorrent/tt-metal/issues/17132
